src/devices/machine/k056230.cpp

```diff
@@ -119,6 +139,210 @@ void k056230_device::ram_w(offs_t offset, u32 data, u32 mem_mask)
+				int recv = read_frame(data_size);
+				while (recv > 0)
+				{
+					// check if valid id
+					u8 idx = m_buffer0[0];
+					if (idx <= 0x0e)
+					{
+						if (idx != m_linkid)
+						{
+							// save message to ram
+							int frame_start = (idx & 0x07) * 0x0100;
+							int frame_size = (idx & 0x08) ? 0x0200 : 0x0100;
+							int frame_offset = 0;
+							u32 frame_data = 0;
+
+							for (int j = 0x00 ; j < frame_size ; j += 4)
+							{
+								frame_offset = (frame_start + j) / 4;
+								frame_data = m_buffer0[4 + j] |  m_buffer0[3 + j] << 8 |  m_buffer0[2 + j] << 16 |  m_buffer0[1 + j] << 24;
+								ram_w(frame_offset, frame_data, 0xffffffff);
+							}
+
+							// forward message to other nodes
+							send_frame(data_size);
+						}
+					}
+
+					// try to read another message
+					recv = read_frame(data_size);
+				}
+			}
+			else if (m_txmode == 0x01)
+			{
+				// send local data to network (once)
+				int frame_start = (m_linkid & 0x07) * 0x100;
+				int frame_size = (m_linkid & 0x08) ? 0x200 : 0x100;
+
+				int frame_offset = 0;
+				u32 frame_data = 0;
+
+				m_buffer0[0] = m_linkid;
+				for (int i = 0x00 ; i < frame_size ; i += 4)
+				{
+					frame_offset = (frame_start + i) / 4;
+					frame_data = ram_r(frame_offset, 0xffffffff);
+					m_buffer0[4 + i] = frame_data & 0xff;
+					m_buffer0[3 + i] = (frame_data >> 8) & 0xff;
+					m_buffer0[2 + i] = (frame_data >> 16) & 0xff;
+					m_buffer0[1 + i] = (frame_data >> 24) & 0xff;
+				}
+
+				send_frame(data_size);
+				m_txmode = 0x02;
+			}
+		}
+	}
+}
+
+int k056230_device::read_frame(int data_size)
+{
+	if (!m_line_rx)
+		return 0;
+
+	// try to read a message (handling partial reads)
+	u32 bytes_total = 0;
+	u32 bytes_read = 0;
+	while (bytes_total < data_size)
+	{
+		std::error_condition filerr = m_line_rx->read(&m_buffer0[bytes_total], 0, data_size - bytes_total, bytes_read);
+		if (filerr)
+		{
+			bytes_read = 0;
+			// special case if first read returned "no data"
+			if (bytes_total == 0 && std::errc::operation_would_block == filerr)
+				return 0;
+		}
+		if ((!filerr && bytes_read == 0) || (filerr && std::errc::operation_would_block != filerr))
+		{
+			osd_printf_verbose("k056230: rx connection failed - %s:%d %s\n", filerr.category().name(), filerr.value(), filerr.message());
+			m_line_rx.reset();
+			return 0;
+		}
+		bytes_total += bytes_read;
+	}
+	return data_size;
+}
+
+void k056230_device::send_frame(int data_size)
+{
+	if (!m_line_tx)
+		return;
+
+	// try to send a message (handling partial writes)
+	u32 bytes_total = 0;
+	u32 bytes_sent = 0;
+	while (bytes_total < data_size)
+	{
+		std::error_condition filerr = m_line_tx->write(&m_buffer0[bytes_total], 0, data_size - bytes_total, bytes_sent);
+		if (filerr)
+		{
+			osd_printf_verbose("k056230: tx connection failed - %s:%d %s\n", filerr.category().name(), filerr.value(), filerr.message());
+			m_line_tx.reset();
+			return;
+		}
+		bytes_total += bytes_sent;
+	}
```

Comment:
As I said on a previous iteration, you really don’t want to be holding up emulation waiting to send data.  You want to defer sending the remaining data until a point where you won’t get blocked.